syncManager.py

```diff
@@ -0,0 +1,65 @@
+import threading
+from threading import Barrier, Event
+
+
+class SyncManager:
+    _instance = None
+    _lock = threading.RLock()
+
+    def __new__(cls, barrier_size):
+        with cls._lock:
+            if cls._instance is None:
+                cls._instance = super(SyncManager, cls).__new__(cls)
+                cls._initialize(barrier_size)
+        return cls._instance
+
+    @staticmethod
+    def _initialize(barrier_size):
+        SyncManager._instance.start_barrier = Barrier(barrier_size)
+        SyncManager._instance.client_finished = Event()
+        SyncManager._instance.server_alive = Event()
+
+    @classmethod
+    def reset(cls, barrier_size):
+        with cls._lock:
+            if barrier_size >= 0:
+                if cls._instance is None:
+                    cls.__new__(cls, barrier_size)
+                else:
+                    cls._initialize(barrier_size)
+
+    @classmethod
+    def client_finished(cls):
+        if cls._instance:
+            return cls._instance.client_finished
+
+    @classmethod
+    def set_client_finished(cls):
+        if cls._instance:
+            cls._instance.client_finished.set()
+
+    @classmethod
+    def set_server_alive(cls):
+        if cls._instance:
+            cls._instance.server_alive.set()
+
+    @classmethod
+    def wait_on_barrier(cls):
```

Comment:
Can we add a comment here explaining the behavior and how threads will decrement the barrier count to ensure they start "together".

Code makes sense to me, but just to make sure we remember how this works in 12 months :)